Добавил viewModel и интерфейсы

Dmitriy-2001 opened the pull request.

Dmitriy-2001 committed:
Добавил viewModel и интерфейсы

Secur1ty-21 commented on app/src/main/java/cy/volleybolley/core/presentation/base/BaseViewModel.kt:
```diff
@@ -0,0 +1,52 @@
+
+    // region Safe Launch
+    protected fun launchSafe(
+        onError: (Throwable) -> Unit = { Log.e(tag, "Error in launchSafe", it) },
```
Лучше тоже пометить как suspend

Secur1ty-21 commented on app/src/main/java/cy/volleybolley/core/presentation/base/BaseViewModel.kt:
```diff
@@ -0,0 +1,52 @@
+        onError: (Throwable) -> Unit = { Log.e(tag, "Error in launchSafe", it) },
+        block: suspend () -> Unit
+    ) {
+        viewModelScope.launch {
```
Тут лучше явно указать Dispatcher.IO, по умолчанию запускается на Main

Secur1ty-21 commented on app/src/main/java/cy/volleybolley/core/presentation/base/BaseViewModel.kt:
```diff
@@ -0,0 +1,52 @@
+    private val _uiState = MutableStateFlow(initialState)
+    val uiState: StateFlow<State> = _uiState.asStateFlow()
+
+    protected fun setState(reducer: State.() -> State) {
+        _uiState.value = _uiState.value.reducer()
+    }
```
У flow есть встроенный потокобезопасный метод обновления состояния (update), поэтому этот метод можно убрать

Secur1ty-21 commented on app/src/main/java/cy/volleybolley/core/presentation/base/MviContract.kt:
```diff
@@ -0,0 +1,16 @@
+package cy.volleybolley.core.presentation.base
+
+/**
+ * Интерфейс состояния экрана
+ */
+interface UiState
+
+/**
+ * События, приходящие из UI
+ */
+interface UiEvent
+
+/**
+ * Одноразовые эффекты (типа Toast, Навигация, Диалог)
+ */
+interface UiEffect
```
Круто придумал с маркерными интерфейсами

Secur1ty-21 commented on app/src/main/java/cy/volleybolley/core/presentation/base/BaseViewModel.kt:
```diff
@@ -0,0 +1,52 @@
+abstract class BaseViewModel<State, Event, UiEffect> : ViewModel() {
+
+    // region Abstract
+    protected abstract val initialState: State
```
Думаю начальное состояние лучше задавать в конструкторе, без обьявления переменной. Или доступ к изначальному состоянию как к переменнной был дан для чего-то специально?

Secur1ty-21 reviewed (changes requested):
Хорошая работа

Dmitriy-2001 committed:
фиксы + VolleyResult

Commit:
#32 Compile and clean fixes
1. Add missing imports for Dispatcher
2. Add suspend keyword in launchSafe onError parameter
3. require in BaseViewModel contract implement MviContract marker interfaces

ShevaVarya commented on app/src/main/java/cy/volleybolley/core/presentation/base/BaseViewModel.kt:
```diff
@@ -0,0 +1,53 @@
+    // region State
+    private val _uiState = MutableStateFlow(initialState)
+    val uiState: StateFlow<State> = _uiState.asStateFlow()
+
```
Немного придираюсь - лишняя строка

ShevaVarya reviewed (commented):
Предложение: добавить документацию в класс, а то лично мне было не совсем понятно для чего методы и поля - может быть это только мне

nemislimus commented on app/src/main/java/cy/volleybolley/core/domain/model/VolleyResult.kt:
```diff
@@ -0,0 +1,27 @@
+package cy.volleybolley.core.domain.model
+
+sealed interface VolleyResult<Data, Error> {
+    val isSuccess: Boolean
+    val isFailure: Boolean
```
Зачем эти поля?😅

Secur1ty-21 replied:
Для возможности проверки результата через if, не всегда результат возвращается с данными и через if может быть удобнее

nemislimus commented on app/src/main/java/cy/volleybolley/core/presentation/base/BaseViewModel.kt:
```diff
@@ -0,0 +1,53 @@
+                onError(e)
+            }
+        }
+    }
```
Тут наверно onError лучше сделать блоком для каких то действий при ошибке? 
Типа того:
```kotlin
protected fun launchSafe(
        dispatcher: CoroutineDispatcher = Dispatchers.IO,
        onError: (() -> Unit)? = null,
        block: suspend () -> Unit
    ) {
        viewModelScope.launch(dispatcher) {
            try {
                block()
            } catch (e: Exception) {
                onError?.let { onError() }
                Log.e(tag, "Error in launchSafe", e)
            }
        }
    }
```

Secur1ty-21 replied:
Тут была идея, что если переопределяем onError, то пишем более кастомную ошибку в место вызова. Закреплю контракт кастомного логирования в функцию, а параметр onError поправлю в соответствии с предложенным

nemislimus commented on app/src/main/java/cy/volleybolley/core/presentation/base/BaseViewModel.kt:
```diff
@@ -0,0 +1,53 @@
+    // endregion
+
+    // region One-time Events
+    private val _effect = Channel<Effect>(Channel.BUFFERED)
+    val uiEffect: Flow<Effect> = _effect.receiveAsFlow()
+
+    protected fun sendUiEffect(effect: Effect) {
+        viewModelScope.launch {
+            _effect.send(effect)
+        }
+    }
```
Прикольная, но незнакомая мне штука. Немного почитал...
Возможно я не прав, но может быть лучше сделать nullable дженерик? Чтобы для разового события первым значением на экране был null...

Secur1ty-21 replied:
Хорошее замечание

nemislimus reviewed (approved):
В целом всё отлично.
Комментарии оставил.

Commit:
#32 Fixes according to pr comment
1. Add documentation to base class
2. Add ability to use nullable type for one event fields and methods
3. Change launceSafe contract for require use custom logging in call place
